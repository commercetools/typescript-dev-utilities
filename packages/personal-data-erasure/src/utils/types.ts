/* Logger */
export type LoggerOptions = {
  error: (...args: Array<unknown>) => void;
  info: (...args: Array<unknown>) => void;
  warn: (...args: Array<unknown>) => void;
  debug: (...args: Array<unknown>) => void;
};

export type ApiConfigOptions = {
  host: string;
  projectKey: string;
  credentials: {
    clientId: string;
    clientSecret: string;
    projectKey?: string;
  };
  scopes?: Array<string>;
  apiUrl?: string;
};

export type ErasureOptions = {
  apiConfig: ApiConfigOptions;
  accessToken?: string;
  logger: LoggerOptions;
};

export type Customer = {
  id: string;
  version: number;
  customerNumber?: string;
  key?: string;
  email: string;
  password: string;
  firstName?: string;
  lastName?: string;
  middleName?: string;
  title?: string;
  salutation?: string;
  dateOfBirth?: string;
  companyName?: string;
  vatId?: string;
  addresses: Array<object>;
  defaultShippingAddressId?: string;
  shippingAddressIds?: Array<string>;
  billingAddressIds?: Array<string>;
  defaultBillingAddressId?: string;
  isEmailVerified: boolean;
  createdAt: string;
  lastModifiedAt: string;
  lastMessageSequenceNumber?: number;
  externalId?: string;
  customerGroup?: object;
  custom?: object;
  locale?: string;
};

export type Order = {
  id: string;
  version: number;
  createdAt: string;
  lastModifiedAt: string;
  completedAt?: string;
  orderNumber?: string;
  customerId?: string;
  customerEmail?: string;
  anonymousId?: string;
  lineItemsId?: string;
  totalPrice: object;
  taxedPrice?: object;
  shippingAddress?: object;
  billingAddress?: object;
  taxMode: object;
  taxRoundingMode: object;
  taxCalculationMode: object;
  customerGroup?: object;
  country?: string;
  orderState: object;
  state?: object;
  shipmentState?: object;
  paymentState?: object;
  shippingInfo?: object;
  syncInfo: Set<object>;
  returnInfo: Set<object>;
  discountCodes: Array<object>;
  lastMessageSequenceNumber: number;
  cart?: object;
  custom?: object;
  paymentInfo?: object;
  locale?: string;
  inventoryMode: object;
  shippingRateInput?: object;
  origin: object;
  itemShippingAddresses: Array<object>;
};

export type Cart = {
  id: string;
  version: number;
  customerId?: string;
  customerEmail?: string;
  anonymousId?: string;
  lineItemsId?: string;
  totalPrice: object;
  taxedPrice?: object;
  cartState: object;
  shippingAddress?: object;
  billingAddress?: object;
  inventoryMode: object;
  taxMode: object;
  taxRoundingMode: object;
  taxCalculationMode: object;
  customerGroup?: object;
  country?: string;
  shippingInfo?: object;
  discountCodes: Array<object>;
  refusedGifts: Array<object>;
  custom?: object;
  paymentInfo?: object;
  locale?: string;
  deleteDaysAfterLastModification?: number;
  shippingRateInput?: object;
  origin: object;
  createdAt: string;
  lastModifiedAt: string;
  itemShippingAddresses: Array<object>;
};

export type Payment = {
  id: string;
  version: number;
  key?: string;
  customer?: object;
  anonymousId?: string;
  interfaceId?: string;
  amountPlanned: object;
  paymentMethodInfo: object;
  paymentStatus: object;
  transactions: Array<object>;
  interfaceInteractions: Array<object>;
  custom?: object;
  createdAt: string;
  lastModifiedAt: string;
};

export type ShoppingList = {
  id: string;
  key?: string;
  version: number;
  createdAt: string;
  lastModifiedAt: string;
  slug?: string;
  name: string;
  description?: string;
  customer?: object;
  anonymousId?: string;
  lineItems: Array<object>;
  textLineItems: Array<object>;
  custom?: object;
  deleteDaysAfterLastModification?: number;
};

export type Review = {
  id: string;
  version: number;
  createdAt: string;
  lastModifiedAt: string;
  key?: string;
  uniquenessValue?: string;
  locale?: string;
  authorName?: string;
  title?: string;
  text?: string;
  target?: object;
  rating?: number;
  state?: object;
  includedInStatistics: boolean;
  customer?: object;
  custom?: object;
};

export type Message = {
  id: string;
  version: number;
  sequenceNumber: number;
  resource: object;
  resourceVersion: number;
  type: string;
  customer?: Customer;
  createdAt: string;
  lastModifiedAt: string;
};

export type Messages = Array<Message>;

export type AllData = Customer | Order | Cart | Payment | ShoppingList | Review;

// sdk
export type HttpHeaders = {
  [key: string]: string;
};

export type SuccessResult<T = object> = {
  body: {
    count: number;
    results: Array<T>;
  };
  statusCode: number;
  headers?: HttpHeaders;
};

// client
export type MethodType =
  | 'GET'
  | 'POST'
  | 'DELETE'
  | 'HEAD'
  | 'OPTIONS'
  | 'PUT'
  | 'PATCH'
  | 'TRACE';

export type ClientRequest = {
  uri: string;
  method: MethodType;
  body?: string | object;
  headers?: HttpHeaders;
};

export type HttpErrorType = {
  name: string;
  message: string;
  code: number;
  status: number;
  statusCode: number;
  originalRequest: ClientRequest;
  body?: object;
  headers?: HttpHeaders;
};

export type ClientResponse = {
  body?: object;
  error?: HttpErrorType;
  statusCode: number;
  headers?: HttpHeaders;
  request?: object;
};

export type ClientResult = SuccessResult | HttpErrorType;
export type ISuccessResponse<T> = {
  limit: number;
  offset: number;
  count: number;
  results: Array<T>;
};

export type CustomClientResult = ClientResult & {
  id: string;
  version?: number;
};

export type DeleterOptions = {
  apiConfig: ApiConfigOptions;
  accessToken?: string;
  predicate?: string;
  logger?: LoggerOptions;
  resource: string;
  fetchedResource: CustomClientResult;
};

export type MethodNames =
  | 'carts'
  | 'categories'
  | 'channels'
  | 'customerGroups'
  | 'customers'
  // | 'customObjects'
  | 'discountCodes'
  | 'inventory'
  | 'orders'
  | 'payments'
  | 'productDiscounts'
  | 'shoppingLists'
  | 'products'
  | 'productTypes'
  | 'reviews'
  | 'shippingMethods'
  | 'states'
  | 'taxCategories'
  | 'types'
  | 'zones';
