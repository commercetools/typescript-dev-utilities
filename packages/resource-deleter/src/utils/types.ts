import { ByProjectKeyRequestBuilder } from '@commercetools/platform-sdk';
import {
  SuccessResult,
  ClientRequest,
  ClientResponse,
  ClientResult,
} from '@commercetools/ts-client';

/* Logger */
export type LoggerOptions = {
  error: (...args: Array<any>) => void;
  info: (...args: Array<any>) => void;
  warn: (...args: Array<any>) => void;
  debug: (...args: Array<any>) => void;
};

export type HttpHeaders = {
  [key: string]: string;
};

// client
export type MethodType =
  | 'GET'
  | 'POST'
  | 'DELETE'
  | 'HEAD'
  | 'OPTIONS'
  | 'PUT'
  | 'PATCH'
  | 'TRACE';

// export type ClientRequest = {
//   uri: string;
//   method: MethodType;
//   body?: string | Object;
//   headers?: HttpHeaders;
// };

export type HttpErrorType = {
  name: string;
  message: string;
  code: number;
  status: number;
  statusCode: number;
  originalRequest: ClientRequest;
  body?: Object;
  headers?: HttpHeaders;
};

// export type SuccessResult<T = object> = {
//   body: {
//     count: number;
//     results: Array<T>;
//   };
//   statusCode: number;
//   headers?: HttpHeaders;
// };

// export type ClientResult = SuccessResult | HttpErrorType;
export type ISuccessResponse<T> = {
  limit: number;
  offset: number;
  count: number;
  results: Array<T>;
};

export type ApiConfigOptions = {
  host: string;
  projectKey: string;
  credentials: {
    clientId: string;
    clientSecret: string;
    projectKey?: string;
  };
  scopes?: Array<string>;
  apiUrl?: string;
};

export type CustomClientResult = ClientResult & {
  id: string;
  version?: number;
};

export type DeleterOptions = {
  apiConfig: ApiConfigOptions;
  accessToken?: string;
  predicate?: string;
  logger?: LoggerOptions;
  resource: MethodNames;
  // fetchedResource: CustomClientResult;
};

export type MethodNames =
  | 'carts'
  | 'categories'
  | 'channels'
  | 'customerGroups'
  | 'customers'
  // | "customObjects"
  | 'discountCodes'
  | 'inventory'
  | 'orders'
  | 'payments'
  | 'productDiscounts'
  | 'products'
  | 'productTypes'
  | 'reviews'
  | 'shippingMethods'
  | 'states'
  | 'taxCategories'
  | 'types'
  | 'zones';

export type InstanceMethods<T> = {
  [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never;
}[keyof T];

export type MethodReturnTypes<T> = {
  [K in InstanceMethods<T>]: T[K] extends (...args: any[]) => infer R
    ? R
    : never;
};
