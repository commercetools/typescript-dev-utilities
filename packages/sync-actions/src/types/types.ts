import createBuildActions from '../utils-ts/create-build-actions';
import createMapActionGroup from '../utils-ts/create-map-action-group';
import * as typeActions from './types-actions';
import * as diffPatcher from '../utils-ts/diffpatcher';
import findMatchingPairs from '../utils-ts/find-matching-pairs';
import {
  ActionGroup,
  Delta,
  SyncAction,
  SyncActionConfig,
  Type,
  TypeUpdateAction,
  UpdateAction,
} from '../utils-ts/types';

const actionGroups = ['base', 'fieldDefinitions'];

function createTypeMapActions(
  mapActionGroup: (
    type: string,
    fn: () => Array<UpdateAction>
  ) => Array<UpdateAction>,
  syncActionConfig: SyncActionConfig
) {
  return function doMapActions(diff: Delta, next: Type & {}, previous) {
    const allActions = [];
    allActions.push(
      mapActionGroup('base', () =>
        typeActions.actionsMapBase(diff, previous, next, syncActionConfig)
      ),
      mapActionGroup('fieldDefinitions', () =>
        typeActions.actionsMapFieldDefinitions(
          diff.fieldDefinitions,
          previous.fieldDefinitions,
          next.fieldDefinitions,
          findMatchingPairs(
            diff.fieldDefinitions,
            previous.fieldDefinitions,
            next.fieldDefinitions,
            'name'
          )
        )
      )
    );
    return allActions.flat();
  };
}

export default (
  actionGroupList: Array<ActionGroup>,
  syncActionConfig: SyncActionConfig
): SyncAction<Type, TypeUpdateAction> => {
  const mapActionGroup = createMapActionGroup(actionGroupList);
  const doMapActions = createTypeMapActions(mapActionGroup, syncActionConfig);
  const buildActions = createBuildActions(diffPatcher.diff, doMapActions);
  return { buildActions };
};

export { actionGroups };
