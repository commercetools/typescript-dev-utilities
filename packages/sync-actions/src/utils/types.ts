import type { Delta as DiffDelta } from 'jsondiffpatch/dist/jsondiffpatch.cjs';

export * from '@commercetools/platform-sdk';
export type Delta = DiffDelta | undefined;

export type SyncActionConfig = {
  shouldOmitEmptyString?: boolean;
  shouldUnsetOmittedProperties?: boolean;
  shouldPreventUnsettingRequiredFields?: boolean;
  sameForAllAttributeNames?: unknown[];
  enableDiscounted?: boolean;
  withHints?: boolean;
  [key: string]: unknown;
};

export type SyncAction<
  R extends object | undefined,
  S extends UpdateAction | unknown,
  T extends object = SyncActionConfig,
> = {
  buildActions: (
    now: DeepPartial<R>,
    before: DeepPartial<R>,
    config?: T
  ) => Array<S>;
};

export type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>;
    }
  : T;

export type UpdateAction = {
  action: string | never;
  key?: string;
  actionKey?: string;
  roles?: Array<unknown>;
  value?: string;
  // [key: string | number | symbol]: unknown;
};

export type ActionGroup = {
  type: string;
  group: 'ignore' | 'allow' | string;
};

export type MapActionGroup = (
  type: string,
  cb: () => Array<UpdateAction>
) => Array<UpdateAction>;

export type MapAction<T> = (
  mapActionGroup: MapActionGroup,
  syncActionConfig?: SyncActionConfig
) => (
  diff: Delta,
  newObj: T,
  oldObj: T,
  config?: SyncActionConfig
) => Array<UpdateAction>;
