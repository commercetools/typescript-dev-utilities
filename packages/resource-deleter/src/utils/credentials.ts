import path from "path";
import dotenv from "dotenv";
import { homepage } from "../../package.json";

export type Credentials = {
  clientId: string;
  clientSecret: string;
  projectKey?: string;
};

export function setCredentialsFromEnvFile() {
  const currentDirectoryResult = dotenv.config({
    path: path.resolve(".ct-credentials.env"),
  });

  const etcDirectoryResult = dotenv.config({
    path: path.resolve(path.join("/etc", ".ct-credentials.env")),
  });

  return {
    ...currentDirectoryResult.parsed,
    ...etcDirectoryResult.parsed,
  };
}

export function getCredentialsFromEnvironment(): Promise<Credentials> {
  return new Promise((resolve, reject) => {
    const clientId = process.env["CTP_CLIENT_ID"];
    const clientSecret = process.env["CTP_CLIENT_SECRET"];
    const projectKey = process.env["CTP_PROJECT_KEY"];

    const credentials = { projectKey, clientId, clientSecret };
    const keys = Object.keys(credentials);

    for (let i = 0; i < keys.length; i++) {
      if (!keys[i]) {
        return reject(
          new Error(`
            Could not find environment variable ${keys[i]}
            see ${homepage}#usage
          `)
        );
      }
    }

    return resolve(credentials);
  });
}

export default async function getCredentials() {
  try {
    await Promise.resolve(setCredentialsFromEnvFile());
    return getCredentialsFromEnvironment();
  } catch (err) {
    return Promise.reject(err);
  }
}
