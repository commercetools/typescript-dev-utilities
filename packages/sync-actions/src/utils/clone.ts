type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

export default function clone<T>(obj: T): Mutable<T> {
  return JSON.parse(JSON.stringify(obj));
}

export function notEmpty<T>(value: T | null | undefined): value is T {
  return value !== null && value !== undefined;
}
