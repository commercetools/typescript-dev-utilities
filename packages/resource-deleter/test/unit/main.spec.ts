require("dotenv").config();
import ResourceDeleter from "../../src/main";
import silentLogger from "../../src/utils/logger";
import { describe, expect, test } from "@jest/globals";
import { MethodNames } from "../../src/utils/types";

describe("ResourceDeleter", () => {
  const logger = {
    ...silentLogger,
  };

  const ID = "sample-ID";
  const options = {
    logger,
    apiConfig: {
      host: "https://sample-host.com",
      apiUrl: "https://sample-api-url.com",
      projectKey: "sample-project-key",
      credentials: {
        clientId: "sample-clientId",
        clientSecret: "sample-clientSecret",
      },
    },
    resource: "categories" as MethodNames,
    predicate: `id=${ID}`,
  };

  let resourceDeleter;
  beforeEach(async () => {
    resourceDeleter = new ResourceDeleter(options);
  });

  describe("::constructor", () => {
    test("should be a function", () => {
      expect(typeof resourceDeleter).toBe("object");
      expect(typeof ResourceDeleter).toBe("function");
    });

    test("should set default properties", () => {
      expect(resourceDeleter.logger).toEqual(logger);
      expect(resourceDeleter.apiConfig).toEqual(options.apiConfig);
      expect(resourceDeleter.resource).toEqual(options.resource);
    });

    test("should if resource property is undefined", () => {
      expect(
        () => new ResourceDeleter({ ...options, resource: undefined })
      ).toThrow("A `resource` string must be passed");
    });

    test("should configure with instance with an existing token", () => {
      const _resourceDeleter = new ResourceDeleter({
        accessToken: "access-token",
        ...options,
      });

      expect(_resourceDeleter).toHaveProperty("logger");
      expect(_resourceDeleter).toHaveProperty("client");
      expect(_resourceDeleter).toHaveProperty("apiRoot");
    });

    test("should throw error if no `apiConfig` in `options` parameter", () => {
      expect(
        () => new ResourceDeleter({ foo: "bar" } as any)
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe("::run", () => {
    let payload;
    describe("with status code 200", () => {
      beforeEach(() => {
        payload = {
          statusCode: 200,
          body: {
            results: [
              { id: "foo1", key: "fooKey", version: 1 },
              { id: "boo2", key: "booKey", version: 2 },
              { id: "fooboo3", key: "foboKey", version: 3 },
            ],
          },
        };

        resourceDeleter.logger.info = jest.fn();
        resourceDeleter.client = {
          execute: jest.fn(() => Promise.resolve(payload)),
          process: jest.fn((_, callback, {}) => {
            callback(payload);
            return Promise.resolve([payload]);
          }),
        };
      });

      test("should delete fetched resource", async () => {
        const noOfResourceToDelete = payload.body.results.length;
        await resourceDeleter.run();
        expect(resourceDeleter.logger.info).toHaveBeenCalledTimes(3);
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          1,
          "Starting to delete resource..."
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          2,
          `Deleting ${noOfResourceToDelete} resources`
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          3,
          `A total of ${noOfResourceToDelete} ${resourceDeleter.resource} have been removed`
        );
      });
    });

    describe("should show message that no resource is found when resource is empty with status code 200", () => {
      beforeEach(() => {
        payload = {
          statusCode: 200,
          body: {
            results: [],
          },
        };

        resourceDeleter.client = {
          execute: jest.fn(() => Promise.resolve(payload)),
          process: jest.fn((_, callback, {}) => {
            callback(payload);
            return Promise.resolve([payload]);
          }),
        };
      });

      test("should resolve for an empty resource", async () => {
        await resourceDeleter.run();
        await expect(Promise.resolve("nothing to delete")).resolves.toBe(
          "nothing to delete"
        );
      });
    });

    describe("should delete product that is published", () => {
      beforeEach(() => {
        payload = {
          statusCode: 200,
          body: {
            results: [
              {
                id: "foo1",
                version: 1,
                masterData: { published: true },
              },
            ],
          },
        };

        resourceDeleter.logger.info = jest.fn();
        resourceDeleter.logger.error = jest.fn();
        resourceDeleter.unPublishResource = jest.fn();
        resourceDeleter.client = {
          execute: jest.fn(() => Promise.resolve(payload)),
          process: jest.fn((_, callback, {}) => {
            callback(payload);
            return Promise.resolve([payload]);
          }),
        };
      });

      test("should delete published resource", async () => {
        const noOfResourceToDelete = payload.body.results.length;
        await resourceDeleter.run();
        expect(resourceDeleter.logger.info).toHaveBeenCalledTimes(3);
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          1,
          "Starting to delete resource..."
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          2,
          `Deleting ${noOfResourceToDelete} resources`
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          3,
          `A total of ${noOfResourceToDelete} ${resourceDeleter.resource} have been removed`
        );
        expect(resourceDeleter.logger.error).not.toHaveBeenCalled();
      });
    });

    describe("should delete categories without children", () => {
      beforeEach(() => {
        resourceDeleter = new ResourceDeleter(options);
        payload = {
          statusCode: 200,
          body: {
            results: [
              {
                id: "barCat",
                version: 1,
                ancestors: [],
              },
            ],
          },
        };

        resourceDeleter.client.process = jest.fn((_, callback, {}) => {
          callback(payload);
          return Promise.resolve([payload]);
        });
        resourceDeleter.client.execute = jest
          .fn(() => Promise.resolve())
          .mockImplementationOnce(() => Promise.resolve(payload))
          .mockImplementationOnce(() => Promise.resolve(payload.body.result[0]))
          .mockImplementationOnce(() => Promise.resolve());

        resourceDeleter.deleteResource = jest.fn();
        resourceDeleter.logger.info = jest.fn();
      });

      test("should delete children categories before deleting the parent", async () => {
        await resourceDeleter.run();
        const noOfResourceToDelete = payload.body.results.length;
        expect(resourceDeleter.deleteResource).toHaveBeenCalledTimes(1);
        expect(resourceDeleter.logger.info).toHaveBeenCalledWith(
          `A total of ${noOfResourceToDelete} ${resourceDeleter.resource} have been removed`
        );
      });
    });

    describe("should delete categories with children", () => {
      beforeEach(() => {
        resourceDeleter = new ResourceDeleter(options);
        payload = {
          statusCode: 200,
          body: {
            results: [
              {
                id: "barParent123",
                version: 1,
                ancestors: [],
              },
              {
                id: "barChild1",
                version: 1,
                ancestors: [{ id: "barParent123", typeId: "category" }],
              },
              {
                id: "barChild2",
                version: 1,
                ancestors: [{ id: "barParent123", typeId: "category" }],
              },
            ],
          },
        };

        resourceDeleter.client = {
          execute: jest.fn(() => Promise.resolve(payload)),
          process: jest.fn((_, callback, {}) => {
            callback(payload);
            return Promise.resolve([payload]);
          }),
        };
        resourceDeleter.client.execute = jest
          .fn(() => Promise.resolve())
          .mockImplementationOnce(() => Promise.resolve(payload))
          .mockImplementationOnce(() => Promise.resolve(payload.body.result[0]))
          .mockImplementationOnce(() => Promise.resolve());

        resourceDeleter.deleteResource = jest.fn();
        resourceDeleter.logger.info = jest.fn();
      });

      test("should delete children categories before deleting the parent", async () => {
        await resourceDeleter.run();
        const noOfResourceToDelete = payload.body.results.length;
        expect(resourceDeleter.deleteResource).toHaveBeenCalledTimes(3);
        expect(resourceDeleter.logger.info).toHaveBeenCalledTimes(3);
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          1,
          "Starting to delete resource..."
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          2,
          `Deleting ${noOfResourceToDelete} resources`
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          3,
          `A total of ${noOfResourceToDelete} ${resourceDeleter.resource} have been removed`
        );
      });
    });

    describe("should delete categories with & without children", () => {
      beforeEach(() => {
        resourceDeleter = new ResourceDeleter(options);
        payload = {
          statusCode: 200,
          body: {
            results: [
              {
                id: "barCat21",
                version: 1,
                ancestors: [],
              },
              {
                id: "fooCat1",
                version: 1,
                ancestors: [],
              },
              {
                id: "fooCat2",
                version: 1,
                ancestors: [],
              },
              {
                id: "barParent123",
                version: 1,
                ancestors: [],
              },
              {
                id: "barChild1",
                version: 1,
                ancestors: [{ id: "barParent123", typeId: "category" }],
              },
              {
                id: "barChild2",
                version: 1,
                ancestors: [{ id: "barParent123", typeId: "category" }],
              },
              {
                id: "barGrandChild21",
                version: 1,
                ancestors: [
                  { id: "barParent123", typeId: "category" },
                  { id: "barChild2", typeId: "category" },
                ],
              },
              {
                id: "barGrandChild22",
                version: 1,
                ancestors: [
                  { id: "barParent123", typeId: "category" },
                  { id: "barChild2", typeId: "category" },
                ],
              },
              {
                id: "barGreatGrandChild21",
                version: 1,
                ancestors: [
                  { id: "barParent123", typeId: "category" },
                  { id: "barChild2", typeId: "category" },
                  { id: "barGrandChild21", typeId: "category" },
                ],
              },
              {
                id: "barGreatGrandChild22",
                version: 1,
                ancestors: [
                  { id: "barParent123", typeId: "category" },
                  { id: "barChild2", typeId: "category" },
                  { id: "barGrandChild21", typeId: "category" },
                ],
              },
              {
                id: "barGGreatGrandChild22",
                version: 1,
                ancestors: [
                  { id: "barParent123", typeId: "category" },
                  { id: "barChild2", typeId: "category" },
                  { id: "barGrandChild21", typeId: "category" },
                  { id: "barGreatGrandChild22", typeId: "category" },
                ],
              },
            ],
          },
        };

        resourceDeleter.client.process = jest.fn((_, callback, {}) => {
          callback(payload);
          return Promise.resolve([payload]);
        });
        resourceDeleter.client.execute = jest
          .fn(() => Promise.resolve())
          .mockImplementationOnce(() => Promise.resolve(payload))
          .mockImplementationOnce(() => Promise.resolve(payload.body.result[0]))
          .mockImplementationOnce(() => Promise.resolve());

        resourceDeleter.deleteResource = jest.fn();
        resourceDeleter.logger.info = jest.fn();
      });

      test("should delete categories without children first before deleting others", async () => {
        const noOfResourceToDelete = payload.body.results.length;
        await resourceDeleter.run();
        expect(resourceDeleter.deleteResource).toHaveBeenCalledTimes(11);
        expect(resourceDeleter.logger.info).toHaveBeenCalledTimes(3);
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          1,
          "Starting to delete resource..."
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          2,
          `Deleting ${noOfResourceToDelete} resources`
        );
        expect(resourceDeleter.logger.info).toHaveBeenNthCalledWith(
          3,
          `A total of ${noOfResourceToDelete} ${resourceDeleter.resource} have been removed`
        );
      });
    });

    describe("should throw error during categories deletion when problem occur", () => {
      beforeEach(() => {
        resourceDeleter = new ResourceDeleter(options);
        payload = {
          statusCode: 200,
          body: {
            results: [
              {
                id: "barParent2",
                version: 1,
                ancestors: [],
              },
              {
                id: "barChild21",
                version: 1,
                ancestors: [{ id: "barParent23", typeId: "category" }],
              },
            ],
          },
        };

        resourceDeleter.client.process = jest.fn((_, callback, {}) => {
          callback(payload);
          return Promise.resolve([payload]);
        });

        resourceDeleter.client.execute = jest
          .fn()
          .mockImplementationOnce(() => Promise.resolve(payload))
          .mockImplementation(() =>
            Promise.reject(new Error("error during `categories` deletion"))
          );
      });

      test("should throw error when there is a problem during categories deletion ", async () => {
        const firstResult = await resourceDeleter.client.execute();
        expect(firstResult).toEqual(payload);

        await expect(resourceDeleter.client.execute()).rejects.toThrow(
          "error during `categories` deletion"
        );
      });
    });

    describe("should throw error when requires parameters are not passed with status code 200", () => {
      beforeEach(() => {
        payload = {
          statusCode: 200,
          body: {
            results: [
              { id: "foo1", key: "fooKey", version: 1 },
              { id: "boo2", key: "booKey", version: 2 },
              { id: "fooboo3", key: "foboKey", version: 3 },
            ],
          },
        };

        resourceDeleter.client.execute = jest
          .fn()
          .mockResolvedValueOnce(payload as never) // First call resolves
          .mockRejectedValue(new Error("error during `resource` deletion"));
      });

      test("should throw error if required parameter are missing with the resource during deletion", async () => {
        const firstResult = await resourceDeleter.client.execute();
        expect(firstResult).toEqual(payload);

        await expect(resourceDeleter.client.execute()).rejects.toThrow(
          "error during `resource` deletion"
        );
      });
    });

    describe("with status code 500", () => {
      beforeEach(() => {
        payload = {
          statusCode: 500,
          body: {
            results: [],
          },
        };
        resourceDeleter.client.process = jest.fn(async (request, callback) => {
          await callback(payload);
        });
      });

      test("should throw internal server error", () =>
        expect(resourceDeleter.run()).rejects.toThrow(
          /Request returned status code 500/
        ));
    });
  });

  describe("::deleteResource non 404 error", () => {
    // let payload;
    beforeEach(() => {
      // payload = {
      //   statusCode: 200,
      //   body: {
      //     results: [
      //       { id: "foo1", key: "fooKey", version: 1 },
      //       { id: "boo2", key: "booKey", version: 2 },
      //       { id: "fooboo3", key: "foboKey", version: 3 },
      //     ],
      //   },
      // };

      resourceDeleter.getServiceRequest = jest.fn(() => ({
        uri: "http://sample-resource.uri.com",
        method: "DELETE",
      }));

      resourceDeleter.client.execute = jest
        .fn()
        .mockRejectedValue(new Error("error during `resource` deletion"));
    });

    test("should reject with error if deletion process completed with errors", () => {
      expect(resourceDeleter.deleteResource()).rejects.toThrow(
        /error during `resource` deletion/
      );
    });
  });

  describe("::deleteResource 404 error", () => {
    let payload;
    beforeEach(() => {
      payload = {
        statusCode: 404,
        body: {
          results: [],
        },
      };

      resourceDeleter.getServiceRequest = jest.fn(() => ({
        uri: "http://sample-resource.uri.com",
        method: "DELETE",
      }));

      resourceDeleter.client.execute = jest.fn().mockRejectedValue(payload);
    });

    test("should resolve with undefine if deletion completed with 404 error", () => {
      expect(resourceDeleter.deleteResource()).resolves.toEqual(undefined);
    });
  });

  describe("::createService", () => {
    let resource;
    beforeEach(() => {
      resource = "carts";
      resourceDeleter.resource = resource;
      resourceDeleter.logger.error = jest.fn();
      resourceDeleter.builder = { carts: "not-a-function" };
    });

    test("should throw error if typeof resource builder is not a function", () => {
      try {
        resourceDeleter.createService();
      } catch (e) {
        expect(resourceDeleter.logger.error).toHaveBeenCalledWith(
          `Method ${resource} does not exist on the provided instance`
        );

        expect(e.message).toEqual(
          `Method ${resource} does not exist on the provided instance`
        );
      }
    });
  });

  describe("::unPublishResource", () => {
    let result;
    beforeEach(() => {
      result = { id: "sample-id", version: 1 };
      resourceDeleter.getServiceRequest = jest.fn(() => ({
        uri: options.apiConfig.apiUrl,
        method: "POST",
      }));

      resourceDeleter.client = {
        execute: jest.fn().mockResolvedValue({
          body: { published: false },
          statusCode: 200,
        }),
      };
    });

    test("should unpublish a resource", async () => {
      const res = await resourceDeleter.unPublishResource(result);
      expect(res.published).toEqual(false);
    });
  });

  // private methods
  describe("::buildRequest", () => {
    test("should build default request", () => {
      // expect(ResourceDeleter.buildRequest("example.com", "GET")).toEqual({
      //   uri: "example.com",
      //   method: "GET",
      // });
      // expect(ResourceDeleter.buildRequest("example.com", "DELETE")).toEqual({
      //   uri: "example.com",
      //   method: "DELETE",
      // });

      // const resourceDeleter = new ResourceDeleter({...options, resource: 'carts'})

      // console.log(resourceDeleter., "<<<");
      const response = { id: "sample-id", version: 1 };
      expect(resourceDeleter.getServiceRequest(response)).toEqual(
        expect.objectContaining({
          method: "GET",
          uri: `/${options.apiConfig.projectKey}/${options.resource}/${response.id}?version=${response.version}`,
        })
      );
    });

    test("should build DELETE request", () => {
      const response = { id: "sample-id", version: 1 };
      expect(resourceDeleter.getServiceRequest(response, true)).toEqual(
        expect.objectContaining({
          method: "DELETE",
          uri: `/${options.apiConfig.projectKey}/${options.resource}/${response.id}?version=${response.version}&dataErasure=true`,
        })
      );
    });
  });

  describe("::buildRequestWithPredicate", () => {
    test("should build request with predicate", () => {
      expect(resourceDeleter.buildRequestWithPredicate()).toEqual(
        expect.objectContaining({
          method: "GET",
          uri: `/${options.apiConfig.projectKey}/${
            options.resource
          }?where=${options.predicate.replace("=", "%3D")}&limit=500`,
        })
      );
    });

    test("should build request without predicate", () => {
      const _resourceDeleter = new ResourceDeleter({
        ...options,
        predicate: undefined,
      });

      expect(_resourceDeleter["buildRequestWithPredicate"]()).toEqual(
        expect.objectContaining({
          method: "GET",
          uri: `/${options.apiConfig.projectKey}/${options.resource}?limit=500`,
        })
      );
    });
  });

  describe("::setPredicate", () => {
    test("should set predicate using the `setPredicate` method", () => {
      const newPredicate = "key=foo";
      expect(resourceDeleter.predicate).toEqual(options.predicate);
      resourceDeleter.setPredicate(newPredicate);
      expect(resourceDeleter.predicate).toEqual(newPredicate);
    });
  });

  describe("::setResource", () => {
    test("should set resource using the `setResource` method", () => {
      const newResource = "taxCategories";
      expect(resourceDeleter.predicate).toEqual(options.predicate);
      resourceDeleter.setResource(newResource);
      expect(resourceDeleter.resource).toEqual(newResource);
    });
  });
});
