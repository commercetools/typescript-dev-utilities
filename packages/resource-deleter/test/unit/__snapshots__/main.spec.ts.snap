// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `ResourceDeleter ::constructor should throw error if no \`apiConfig\` in \`options\` parameter 1`
] = `"The constructor must passed an \`apiConfig\` object"`;
